pygfx/utils/compgeo.py

```diff
@@ -0,0 +1,150 @@
+"""Computational geometry.
+
+This may at some pointe be moved to https://github.com/pygfx/pycompgeo.
+"""
+
+
+def get_visible_part_of_line_ndc(ndc1, ndc2):
+    """Get the visible part of the line, given by two homogeneous ndc coords.
+
+    Returns (t1, t2) representing the visible line. If the two values
+    are equal, the line is not in the viewport.
+    """
+
+    # Get closest t for all 4 edges
+    tx1 = binary_search_for_ndc_edge(ndc1, ndc2, -1, 0)
+    tx2 = binary_search_for_ndc_edge(ndc1, ndc2, +1, 0)
+    ty1 = binary_search_for_ndc_edge(ndc1, ndc2, -1, 1)
+    ty2 = binary_search_for_ndc_edge(ndc1, ndc2, +1, 1)
+
+    # Sort (because line can be oriented in any way)
+    tx1, tx2 = min(tx1, tx2), max(tx1, tx2)
+    ty1, ty2 = min(ty1, ty2), max(ty1, ty2)
+
+    # Combine dimensions
+    t1, t2 = max(tx1, ty1), min(tx2, ty2)
+
+    # Sort again
+    t1, t2 = min(t1, t2), max(t1, t2)
+
+    return t1, t2
+
+
+def binary_search_for_ndc_edge(ndc1, ndc2, ref, dim, *, n_iters=10):
+    """Get the t that results in the smallest distance to ref in dimension dim.
+
+    The ndc1 and ndc2 arguments must be in homogeneous ndc coords, so
+    that they can be interpolated. The resulting t represents the point
+    on the line between ndc1 and ncd2.
+
+    This algorithm operates either over x or y, depending on dim (0 or1).
+    So in order to determine the visible piece of the line, this
+    function must be called 4 times.
+    """
+
+    # This algorithm performs a binary search, so that we can cover a
+    # fine grid without doing an exhaustive search.
+    #
+    # Note that (with a perspective camera) the distance to the
+    # reference does not scale linear with t. This means that if we
+    # sample two points, we cannot simply select the half that has the
+    # smallest distance, since the minimum can still be in the other
+    # half. To work around this, we sample 5 locations, and select a
+    # sub-region based on the point with smallest distance. When that
+    # point is on the edge (0 or 4), the subregion is just 25% of the
+    # previous. Otherwise (1, 2, 3) it is 50%.
+    #
+    #  |------|------|------|------|
+    #  0      1      2      3      4
+    #
+    # AK: I spent some time optimizing this code. Using numpy arrays
+    # and np.argmin is not faster. What does help is re-using the sample
+    # positions. With 5 samples per level, we can re-use 2 or 3 of the
+    # samples at each step. By using tuples, copying these into the new
+    # slot is efficient.
```

Comment:
it's a long shot, but perhaps the built in [bisect](https://docs.python.org/3/library/bisect.html) module could be leveraged here as well? not sure, I didn't read the code too carefully

Reply:
Thanks for the suggestion! Looking into it ...

Reply:
Awesome!